### PR TITLE
import: hash URL in paths if they are too long

### PR DESCRIPTION
**Pull request by jsynacek**
https://bugzilla.redhat.com/show_bug.cgi?id=1266775

### CODE REVIEW

File: src/import/pull-common.c [hunk 1]
```diff
@@ -149,8 +150,16 @@ int pull_make_local_copy(const char *final, const char *image_root, const char *
         return 0;
 }
 
+static int hash_url(const char *url, char **ret) {
+        uint64_t h;
+        const uint8_t k[16] = {'s', 'e', 'c', 'r', 'e', 't'};
+
+        siphash24((uint8_t *) &h, url, strlen(url), k);
+        return asprintf(ret, "%lx", h);
```
Comment (zonque) [inline src/import/pull-common.c:158]: `PRIu64`

File: src/import/pull-common.c [hunk 2]
```diff
@@ -164,18 +173,35 @@ int pull_make_path(const char *url, const char *etag, const char *image_root, co
...
-        } else
-                path = strjoin(image_root, "/", strempty(prefix), escaped_url, strempty(suffix), NULL);
+        path = strjoin(image_root, "/", strempty(prefix), escaped_url, escaped_etag ? "." : "",
+                       strempty(escaped_etag), strempty(suffix), NULL);
         if (!path)
                 return -ENOMEM;
 
+        /* URLs might make the path longer than the maximum allowed length for a file name.
+         * When that happens, a URL hash is used instead. Paths returned by this function
+         * can be later used with tempfn_random() which adds 16 bytes to the resulting name. */
```
Comment (zonque) [inline src/import/pull-common.c:188]: Why not always do that, unconditionally? I think having this distinction will cause confusion sooner or later.

Comment (jsynacek) [inline src/import/pull-common.c:188]: Quoting @poettering from PR #1413:
"That way, things stay nicely readable in most cases, and in the cases where we hit the limit things still work, and the names are pretty much always unique."
I agree that hashing it right away is a better option.

File: src/import/pull-common.c [hunk 3]
```diff
@@ -149,8 +150,16 @@ int pull_make_local_copy(const char *final, const char *image_root, const char *
         return 0;
 }
 
+static int hash_url(const char *url, char **ret) {
+        uint64_t h;
+        const uint8_t k[16] = {'s', 'e', 'c', 'r', 'e', 't'};
+
```
Comment (zonque) [inline src/import/pull-common.c:156]: While you're at it: `assert(url);`

**Comment by jsynacek** [comment]
v4: https://github.com/systemd/systemd/pull/1479

**Pull request closed without merging**